### PR TITLE
Vec2D

### PR DESCRIPTION
**Pull request by randaz81**
New class Vec2D is a template class which can be specialized only as int or double.
It contains two elements x,y. Basic math operations as well as conversion operators to yarp::sig::Vector are supported.

### CODE REVIEW

File: src/libYARP_math/include/yarp/math/Vec2D.h [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    explicit operator yarp::sig::Vector() const
+    {
+        yarp::sig::Vector v(2);
+        v[0] = double(x);
+        v[1] = double(y);
+        return v;
+    }
+
+    /**
+    * Returns the modulus of the Vec2D, i.e. srqrt(x^2+y^2)
```
Comment (traversaro) [inline src/libYARP_math/include/yarp/math/Vec2D.h]: Typo : srqrt -> sqrt .

File: src/libYARP_math/include/yarp/math/Vec2D.h [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    * @return the computed modulus
+    */
+    T mod() const;
```
Comment (traversaro) [inline src/libYARP_math/include/yarp/math/Vec2D.h]: I am afraid that `mod` could be ambiguous as a term, check for example: https://it.mathworks.com/help/matlab/ref/mod.html . 
An alternative name for this method that is used both in Matlab, Eigen and YARP itself (http://www.yarp.it/namespaceyarp_1_1math.html#a378045a0a2b0cc5e7d9c6e54709ba127) is `norm`.

**Comment by traversaro** [comment]
If we want to hide implementation from the template (given that no further instantiations are possible for the template), I think with C++11 we can, with [explicit template instantiation](http://en.cppreference.com/w/cpp/language/class_template#Explicit_instantiation). However probably there are some tricky point in how these interacts with exporting symbols in Windows, so I think it is also fine to just keep it as it is.

**Comment by randaz81** [comment]
Thank you, I never tried the ` extern template`, I'll check it!
Btw, do you think we can use some `rvalue` stuff to optimize the class?

**Comment by traversaro** [comment]
I am not sure I understand the question on `rvalue`. However given that the class attributes are just two doubles, I think the default copy-constructor and move-constructor are already as optimized as possible.

**Comment by francesco-romano** [comment]
A part from the inline comments, I don't know if, given that we allow only `int` and `double` specialisations we can move the implementation in the cpp file. Maybe @traversaro knows better if it is possible.

**Comment by traversaro** [comment]
@francesco-romano https://github.com/robotology/yarp/pull/1048#issuecomment-269995903

**Comment by traversaro** [comment]
Also relevant: 
 * http://drake.mit.edu/cxx_inl.html
 * https://github.com/robotology/yarp/pull/454#issuecomment-85742801

**Comment by francesco-romano** [comment]
@traversaro yes, I remember the issues in the LockGuard. But almost two years passed, and I don't know the situation on VS now.

**Comment by randaz81** [comment]
The implementation is now hidden in the .cpp file.
Test passed on both Linux and Win 
@traversaro @francesco-romano @drdanz

**Pull request merged**